cmake/modules/FindFlatBuffers.cmake

```diff
@@ -16,44 +16,82 @@
+  endif()
 
-find_program(FLATBUFFERS_FLATC_EXECUTABLE NAMES flatc)
-find_path(FLATBUFFERS_INCLUDE_DIR NAMES flatbuffers/flatbuffers.h)
+  set(FLATBUFFERS_FLATC_EXECUTABLE ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/bin/flatc)
+  set(FLATBUFFERS_INCLUDE_DIR ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/include)
+
+  externalproject_add(flatbuffers
+                      URL ${FLATBUFFERS_URL}
+                      DOWNLOAD_DIR ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/download
+                      PREFIX ${CORE_BUILD_DIR}/flatbuffers
+                      CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}
+                                 -DCMAKE_BUILD_TYPE=Release
+                                 -DFLATBUFFERS_CODE_COVERAGE=OFF
+                                 -DFLATBUFFERS_BUILD_TESTS=OFF
+                                 -DFLATBUFFERS_INSTALL=ON
+                                 -DFLATBUFFERS_BUILD_FLATLIB=OFF
+                                 -DFLATBUFFERS_BUILD_FLATC=ON
+                                 -DFLATBUFFERS_BUILD_FLATHASH=OFF
+                                 -DFLATBUFFERS_BUILD_GRPCTEST=OFF
+                                 -DFLATBUFFERS_BUILD_SHAREDLIB=OFF
+                                 "${EXTRA_ARGS}"
+                      BUILD_BYPRODUCTS ${FLATBUFFERS_FLATC_EXECUTABLE})
+  set_target_properties(flatbuffers PROPERTIES FOLDER "External Projects")
+else()
+  find_program(FLATBUFFERS_FLATC_EXECUTABLE NAMES flatc)
+  find_path(FLATBUFFERS_INCLUDE_DIR NAMES flatbuffers/flatbuffers.h)
+endif()
 
 include(FindPackageHandleStandardArgs)
 find_package_handle_standard_args(flatbuffers
   DEFAULT_MSG FLATBUFFERS_FLATC_EXECUTABLE FLATBUFFERS_INCLUDE_DIR)
 
-if(FLATBUFFERS_FOUND)
-  function(FLATBUFFERS_GENERATE_C_HEADERS Name)
-    set(FLATC_OUTPUTS)
-    foreach(FILE ${ARGN})
-      get_filename_component(FLATC_OUTPUT ${FILE} NAME_WE)
-      set(FLATC_OUTPUT
-        "${CMAKE_CURRENT_BINARY_DIR}/${FLATC_OUTPUT}_generated.h")
-      list(APPEND FLATC_OUTPUTS ${FLATC_OUTPUT})
-
-      add_custom_command(OUTPUT ${FLATC_OUTPUT}
-        COMMAND ${FLATBUFFERS_FLATC_EXECUTABLE}
-        ARGS -c -o "${CMAKE_CURRENT_BINARY_DIR}/" ${FILE}
-        DEPENDS ${FILE}
-        COMMENT "Building C++ header for ${FILE}"
-        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
-    endforeach()
-    set(${Name}_OUTPUTS ${FLATC_OUTPUTS} PARENT_SCOPE)
-  endfunction()
-
-  set(FLATBUFFERS_INCLUDE_DIRS ${FLATBUFFERS_INCLUDE_DIR})
-  include_directories(${CMAKE_BINARY_DIR})
-else()
-  set(FLATBUFFERS_INCLUDE_DIR)
-endif()
+function(FLATBUFFERS_GENERATE_C_HEADERS Name)
+  set(FLATC_OUTPUTS)
+  foreach(FILE ${ARGN})
+    get_filename_component(FLATC_OUTPUT ${FILE} NAME_WE)
+    set(FLATC_OUTPUT
+      "${CMAKE_CURRENT_BINARY_DIR}/${FLATC_OUTPUT}_generated.h")
+    list(APPEND FLATC_OUTPUTS ${FLATC_OUTPUT})
+
+    add_custom_command(OUTPUT ${FLATC_OUTPUT}
+      COMMAND ${FLATBUFFERS_FLATC_EXECUTABLE}
```

Comment:
@Rechi when this macro is run, `FLATBUFFERS_FLATC_EXECUTABLE` is empty. How can I propagate its value to this function?